Error formatting type

This template is a guideline, not a strict requirement.
- **Version**: 2.4.1 (2.3.2 did not cause the error)
- **Integration**: Command Line
- **Configuration**:
```
version = 2.4.1
style = defaultWithAlign
docstrings = JavaDoc
assumeStandardLibraryStripMargin = true
project.excludeFilters = [
  scalalib/
]
project.git = true
# changed 1.6 or after
align.openParenCallSite = true
align.openParenDefnSite = true
# avoid wrapping parens on new line
danglingParentheses.defnSite = false
danglingParentheses.callSite = false
# leave literals alone
literals.long = unchanged
literals.float = unchanged
literals.double = unchanged
```

## Steps

When running `$ scripts/scalafmt unit-tests/src/test/scala/scala/scalanative/unsafe/TagSuite.scala`
Given code like this:
```scala
 type iovec = CStruct2[Ptr[Byte], // iov_base
                        CSize] // iov_len
```

## Problem
Scalafmt errors
```scala
org.scalafmt.cli.FailedToFormat: /Users/eric/workspace/scala-native/unit-tests/src/test/scala/scala/scalanative/unsafe/TagSuite.scala
Caused by: org.scalafmt.Error$PreciseIncomplete: /Users/eric/workspace/scala-native/unit-tests/src/test/scala/scala/scalanative/unsafe/TagSuite.scala:144: error: Unable to format file due to bug in scalafmt
  type iovec = CStruct2[Ptr[Byte], // iov_base
                                   ^
	at org.scalafmt.Scalafmt$.formatCode(Scalafmt.scala:86)
	at org.scalafmt.cli.ScalafmtCoreRunner$.unsafeHandleFile(ScalafmtCoreRunner.scala:87)
	at org.scalafmt.cli.ScalafmtCoreRunner$.handleFile(ScalafmtCoreRunner.scala:66)
	at org.scalafmt.cli.ScalafmtCoreRunner$.$anonfun$run$3(ScalafmtCoreRunner.scala:44)
	at org.scalafmt.cli.ScalafmtCoreRunner$.$anonfun$run$3$adapted(ScalafmtCoreRunner.scala:43)
	at scala.collection.Iterator.foreach(Iterator.scala:941)
	at scala.collection.Iterator.foreach$(Iterator.scala:941)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1429)
	at scala.collection.parallel.ParIterableLike$Foreach.leaf(ParIterableLike.scala:974)
	at scala.collection.parallel.Task.$anonfun$tryLeaf$1(Tasks.scala:53)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
	at scala.util.control.Breaks$$anon$1.catchBreak(Breaks.scala:67)
	at scala.collection.parallel.Task.tryLeaf(Tasks.scala:56)
	at scala.collection.parallel.Task.tryLeaf$(Tasks.scala:50)
	at scala.collection.parallel.ParIterableLike$Foreach.tryLeaf(ParIterableLike.scala:971)
	at scala.collection.parallel.AdaptiveWorkStealingTasks$WrappedTask.compute(Tasks.scala:153)
	at scala.collection.parallel.AdaptiveWorkStealingTasks$WrappedTask.compute$(Tasks.scala:149)
	at scala.collection.parallel.AdaptiveWorkStealingForkJoinTasks$WrappedTask.compute(Tasks.scala:440)
	at java.util.concurrent.RecursiveAction.exec(RecursiveAction.java:189)
	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
	at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157)

```
Scalafmt formats code like this:
```scala
No code change
```

## Expectation

No errors.

I would like the formatted output to look like this:
```Scala
type iovec = CStruct2[Ptr[Byte], // iov_base
                      CSize]     // iov_len
```

## Workaround

I've found that by writing the code like this the error goes away but no formatting occurs.
```Scala
type iovec = CStruct2[Ptr[Byte], // iov_base
                      CSize     // iov_len
]
Edit: this is not really a workaround but may hint at the problem.
```

--- a/scalafmt-core/shared/src/main/scala/org/scalafmt/internal/BestFirstSearch.scala
+++ b/scalafmt-core/shared/src/main/scala/org/scalafmt/internal/BestFirstSearch.scala
@@ -47,7 +47,7 @@ private class BestFirstSearch private (
   var deepestYet = State.start
   val best = mutable.Map.empty[Int, State]
   val visits = new Array[Int](tokens.length)
-  val keepSlowStates = !pruneSlowStates
+  var keepSlowStates = !pruneSlowStates
 
   type StateHash = Long
 
@@ -247,7 +247,16 @@ private class BestFirstSearch private (
   }
 
   def getBestPath: SearchResult = {
-    val state = shortestPath(State.start, tree.tokens.last)
+    val state = {
+      def run = shortestPath(State.start, tree.tokens.last)
+      val state = run
+      if (null != state) state
+      else {
+        best.clear()
+        keepSlowStates = true
+        run
+      }
+    }
     if (null != state) {
       runner.event(CompleteFormat(explored, state))
       SearchResult(state, reachedEOF = true)

--- a/scalafmt-core/shared/src/main/scala/org/scalafmt/internal/BestFirstSearch.scala
+++ b/scalafmt-core/shared/src/main/scala/org/scalafmt/internal/BestFirstSearch.scala
@@ -45,10 +45,9 @@ private class BestFirstSearch private (
   val noOptimizations = noOptimizationZones(tree)
   var explored = 0
   var deepestYet = State.start
-  var statementCount = 0
   val best = mutable.Map.empty[Int, State]
-  var pathologicalEscapes = 0
   val visits = new Array[Int](tokens.length)
+  val keepSlowStates = !pruneSlowStates
 
   type StateHash = Long
 
@@ -60,17 +59,11 @@ private class BestFirstSearch private (
   /**
     * Returns true if it's OK to skip over state.
     */
-  def shouldEnterState(curr: State): Boolean = {
-    val splitToken = tokens(curr.depth)
-    val insideOptimizationZone =
-      curr.policy.noDequeue || isInsideNoOptZone(splitToken)
-    def hasBestSolution = !pruneSlowStates || insideOptimizationZone || {
+  def shouldEnterState(curr: State): Boolean =
+    keepSlowStates || curr.policy.noDequeue ||
+      isInsideNoOptZone(tokens(curr.depth)) ||
       // TODO(olafur) document why/how this optimization works.
-      val result = !best.get(curr.depth).exists(_.alwaysBetter(curr))
-      result
-    }
-    hasBestSolution
-  }
+      !best.get(curr.depth).exists(_.alwaysBetter(curr))
 
   def shouldRecurseOnBlock(
       ft: FormatToken,
@@ -208,7 +201,7 @@ private class BestFirstSearch private (
           var optimalNotFound = true
           actualSplit.foreach { split =>
             val nextState = curr.next(style, split, splitToken)
-            if (depth == 0 && split.modification.isNewline &&
+            if (!keepSlowStates && depth == 0 && split.modification.isNewline &&
               !best.contains(curr.depth)) {
               best.update(curr.depth, nextState)
             }

--- a/scalafmt-tests/src/test/resources/align/DefaultWithAlign.stat
+++ b/scalafmt-tests/src/test/resources/align/DefaultWithAlign.stat
@@ -487,19 +487,28 @@ object x {
 type iovec = CStruct2[Ptr[Byte], // iov_base
                         CSize] // iov_len
 >>>
-Unable to format file due to bug in scalafmt
+type iovec = CStruct2[
+  Ptr[Byte], // iov_base
+  CSize
+] // iov_len
 <<< #1720 with val
 val iovec = CStruct2[Ptr[Byte], // iov_base
                         CSize](a) // iov_len
 >>>
-Unable to format file due to bug in scalafmt
+val iovec = CStruct2[
+  Ptr[Byte], // iov_base
+  CSize
+](a) // iov_len
 <<< #1720 no align
 align = none
 ===
 type iovec = CStruct2[Ptr[Byte], // iov_base
                         CSize] // iov_len
 >>>
-Unable to format file due to bug in scalafmt
+type iovec = CStruct2[
+  Ptr[Byte], // iov_base
+  CSize
+] // iov_len
 <<< #1720 second SL comment
 type iovec = CStruct2[Ptr[Byte],
                         CSize] // iov_len
@@ -509,7 +518,10 @@ type iovec = CStruct2[Ptr[Byte], CSize] // iov_len
 type iovec = CStruct2[Ptr[Byte], // iov_base
                         CSize]
 >>>
-Unable to format file due to bug in scalafmt
+type iovec = CStruct2[
+  Ptr[Byte], // iov_base
+  CSize
+]
 <<< #1720 nested multi-arg type with two SL comments
 type iovec = CStruct2[Ptr[Byte, // iov_base
                         CSize]] // iov_len
@@ -529,7 +541,10 @@ type iovec = CStruct2[
 type iovec = CStruct2[Ptr[Byte], /* iov_base */
                         CSize]
 >>>
-Unable to format file due to bug in scalafmt
+type iovec = CStruct2[
+  Ptr[Byte], /* iov_base */
+  CSize
+]
 <<< #1720 first inline comment with apply
 val iovec = CStruct2(Ptr(Byte), /* iov_base */
                         CSize)

--- a/scalafmt-tests/src/test/resources/align/DefaultWithAlign.stat
+++ b/scalafmt-tests/src/test/resources/align/DefaultWithAlign.stat
@@ -483,3 +483,58 @@ object x {
   // comment
   println("foo") // comment
 }
+<<< #1720 original, two SL comments
+type iovec = CStruct2[Ptr[Byte], // iov_base
+                        CSize] // iov_len
+>>>
+Unable to format file due to bug in scalafmt
+<<< #1720 with val
+val iovec = CStruct2[Ptr[Byte], // iov_base
+                        CSize](a) // iov_len
+>>>
+Unable to format file due to bug in scalafmt
+<<< #1720 no align
+align = none
+===
+type iovec = CStruct2[Ptr[Byte], // iov_base
+                        CSize] // iov_len
+>>>
+Unable to format file due to bug in scalafmt
+<<< #1720 second SL comment
+type iovec = CStruct2[Ptr[Byte],
+                        CSize] // iov_len
+>>>
+type iovec = CStruct2[Ptr[Byte], CSize] // iov_len
+<<< #1720 first SL comment
+type iovec = CStruct2[Ptr[Byte], // iov_base
+                        CSize]
+>>>
+Unable to format file due to bug in scalafmt
+<<< #1720 nested multi-arg type with two SL comments
+type iovec = CStruct2[Ptr[Byte, // iov_base
+                        CSize]] // iov_len
+>>>
+type iovec = CStruct2[Ptr[
+  Byte, // iov_base
+  CSize
+]] // iov_len
+<<< #1720 one arg with SL comment
+type iovec = CStruct2[Ptr[Byte] // iov_base
+                        ]
+>>>
+type iovec = CStruct2[
+  Ptr[Byte] // iov_base
+]
+<<< #1720 first inline comment
+type iovec = CStruct2[Ptr[Byte], /* iov_base */
+                        CSize]
+>>>
+Unable to format file due to bug in scalafmt
+<<< #1720 first inline comment with apply
+val iovec = CStruct2(Ptr(Byte), /* iov_base */
+                        CSize)
+>>>
+val iovec = CStruct2(
+  Ptr(Byte), /* iov_base */
+  CSize
+)

